src/EditorFeatures/Core/Implementation/SmartIndent/AbstractIndentationService.cs

```diff
@@ -50,53 +50,39 @@ private IEnumerable<IFormattingRule> GetFormattingRules(Document document, int p
             // the next line).  If we're in the latter case, we defer to the Formatting engine
             // as we need it to use all its rules to determine where the appropriate location is
             // for the following tokens to go.
-            if (ShouldUseSmartTokenFormatterInsteadOfIndenter(formattingRules, root, lineToBeIndented, documentOptions, cancellationToken))
+            if (indenter.ShouldUseFormatterIfAvailable())
             {
                 return null;
             }
 
-            return GetBlankLineIndentation(document, lineNumber, indentStyle, cancellationToken);
+            return indenter.GetDesiredIndentation(indentStyle);
         }
 
-        /// <summary>
-        /// Determines indentation for a blank line (i.e. after hitting enter at the end of a line,
-        /// or after moving to a blank line). This indent style cannot be <see cref="FormattingOptions.IndentStyle.None"/>;
-        /// </summary>
         public IndentationResult GetBlankLineIndentation(
             Document document, int lineNumber, FormattingOptions.IndentStyle indentStyle, CancellationToken cancellationToken)
         {
-            if (indentStyle == FormattingOptions.IndentStyle.None)
-            {
-                throw new ArgumentException(nameof(indentStyle));
-            }
-
-            GetIndenterData(document, lineNumber,
-                out var documentOptions, out var root, out var lineToBeIndented,
-                out var formattingRules, cancellationToken);
-
-            var indenter = GetIndenter(
-                document.GetLanguageService<ISyntaxFactsService>(),
-                root.SyntaxTree, lineToBeIndented, formattingRules,
-                documentOptions, cancellationToken);
-
+            var indenter = GetIndenter(document, lineNumber, cancellationToken);
             return indenter.GetDesiredIndentation(indentStyle);
         }
 
-        private void GetIndenterData(Document document, int lineNumber, out DocumentOptionSet documentOptions, out SyntaxNode root, out TextLine lineToBeIndented, out IEnumerable<IFormattingRule> formattingRules, CancellationToken cancellationToken)
+        private AbstractIndenter GetIndenter(Document document, int lineNumber, CancellationToken cancellationToken)
         {
-            documentOptions = document.GetOptionsAsync(cancellationToken).WaitAndGetResult(cancellationToken);
-            root = document.GetSyntaxRootSynchronously(cancellationToken);
+            var documentOptions = document.GetOptionsAsync(cancellationToken).WaitAndGetResult(cancellationToken);
```

Comment:
❗️ If this can be called on a background thread (e.g. from a code fix), it needs to use `WaitAndGetResult_CanCallOnBackground`. Please confirm that the current wait operation is the expected one at this location.

Reply:
Yeah.  In the codefix branch i use WaitAndGetResult_CanCallOnBackground.  That can either come in from the other PR, or i can update this one.  Your call.

Reply:
It's fine to have it in the other. Thanks!